src/NuGet.Services.Metadata.Catalog.Monitoring/Validation/Test/Validator.cs

```diff
@@ -0,0 +1,134 @@
+            ILogger<Validator> logger)
+        {
+            _timestampMetadataResourceV2 = feedToSource[FeedType.HttpV2].GetResource<IPackageTimestampMetadataResource>();
+            Logger = logger;
+            CommonLogger = logger.AsCommon();
+        }
+
+        protected readonly ILogger<Validator> Logger;
+        protected readonly Common.ILogger CommonLogger;
+
+        private readonly IPackageTimestampMetadataResource _timestampMetadataResourceV2;
+
+        public async Task<ValidationResult> Validate(ValidationContext data)
+        {
+            var validationResult = new ValidationResult
+            {
+                Validator = this,
+                Result = TestResult.Pass
+            };
+            
+            try
+            {
+                var shouldRun = await ShouldRun(data);
+                if (shouldRun)
+                {
+                    await RunInternal(data);
+                }
+                else
+                {
+                    validationResult.Result = TestResult.Skip;
+                }
+            }
+            catch (Exception e)
+            {
+                validationResult.Result = TestResult.Fail;
+                validationResult.Exception = e;
+            }
+
+            return validationResult;
+        }
+
+        /// <summary>
+        /// Checks that the current batch of catalog entries contains the entry that was created from the current state of the V2 feed.
+        /// </summary>
+        protected virtual async Task<bool> ShouldRun(ValidationContext data)
+        {
+            var timestampV2 = await _timestampMetadataResourceV2.GetAsync(data);
+            var timestampCatalog = await PackageTimestampMetadata.FromCatalogEntries(data.Client, data.Entries);
+            
+            if (!timestampV2.Last.HasValue)
+            {
+                throw new ValidationTestUtility.ValidationException("Cannot get timestamp data for package from the V2 feed!");
+            }
+
+            if (!timestampCatalog.Last.HasValue)
+            {
+                throw new ValidationTestUtility.ValidationException("Cannot get timestamp data for package from the catalog!");
+            }
+
+            if (timestampCatalog.Last > timestampV2.Last)
+            {
+                throw new ValidationTestUtility.ValidationException("The timestamp in the catalog is newer than the timestamp in the feed! This should never happen because all data flows from the catalog into the feed!");
+            }
+
+            // If the timestamp metadata in the catalog is LESS than that of the feed, we must not be looking at the latest entry that corresponds with this package, so skip the test for now.
+            // If the timestamp metadata in the catalog is EQUAL to that of the feed, we are looking at the latest catalog entry that corresponds with this package, so run the test.
+            return timestampCatalog.Last == timestampV2.Last;
```

Comment:
recommenting what I commented on that comment:

In my tests, this hasn't been a problem. The feed and the catalog carry the same precision. Do you have a suggestion on a better way to do this comparison? I'm concerned that performing a less precise comparison could lead to equally strange issues.